pkg/accumulator/accumulator.go

```diff
@@ -0,0 +1,256 @@
+		firstMapKeysOfAccumulatorData: make([]string, ACCUMULATOR_SIZE),
+		mainDataChannel:               make(chan *evData.EventData),
+		containersData: containersAccumulator{
+			accumulatorDataPerContainer: make(map[string]chan evData.EventData),
+		},
+	}
+
+	return accumulatorSingelton
+}
+
+func GetAccumulator() *Accumulator {
+	if accumulatorSingelton == nil {
+		lock.Lock()
+		defer lock.Unlock()
+		if accumulatorSingelton == nil {
+			logger.L().Debug("Creating accumulatorSingelton now.")
+			accumulatorSingelton = createAccumulator()
+		} else {
+			logger.L().Debug("accumulatorSingelton already created.")
+		}
+	} else {
+		logger.L().Debug("accumulatorSingelton already created.")
+	}
+
+	return accumulatorSingelton
+}
+
+func CreateContainerAccumulator(containerID string, dataChannel chan evData.EventData) *ContainerAccumulator {
+	return &ContainerAccumulator{
+		dataChannel: dataChannel,
+		containerID: containerID,
+	}
+}
+
+func (acc *Accumulator) createNewMap(event *evData.EventData, index int) {
+	slice := make([]evData.EventData, 0)
+	m := make(map[string][]evData.EventData)
+	m[event.GetEventContainerID()] = slice
+	acc.accumulatorData[index] = m
+	acc.firstMapKeysOfAccumulatorData[index] = event.GetEventContainerID()
+}
+
+func (acc *Accumulator) getFirstTimestamp() (time.Time, error) {
+	if len(acc.accumulatorData) < 1 {
+		return time.Time{}, fmt.Errorf("getFirstTimestamp failed the slice accumulatorData has no members")
+	}
+	if len(acc.firstMapKeysOfAccumulatorData) < 1 {
+		return time.Time{}, fmt.Errorf("getFirstTimestamp failed the slice firstMapKeysOfAccumulatorData has no members")
+	}
+	if len(acc.accumulatorData[0][acc.firstMapKeysOfAccumulatorData[0]]) < 1 {
+		return time.Time{}, fmt.Errorf("getFirstTimestamp failed the slice acc.accumulatorData[0][acc.firstMapKeysOfAccumulatorData[0] has no members")
+	}
+	return acc.accumulatorData[0][acc.firstMapKeysOfAccumulatorData[0]][0].GetEventTimestamp(), nil
+}
+
+func (acc *Accumulator) findIndexByTimestampWhenAccumulatorDataIsFull(event *evData.EventData) int {
+	index := 0
+	minTimestamp, err := acc.getFirstTimestamp()
+	if err != nil {
+		logger.L().Warning("findIndexByTimestampWhenAccumulatorDataIsFull fail to find the place to insert the event, fail with error", helpers.Error(err))
+		return -1
+	}
+	for i := range acc.accumulatorData {
+		if i == 0 {
+			continue
+		}
+		if acc.accumulatorData[i][acc.firstMapKeysOfAccumulatorData[i]][0].GetEventTimestamp().Before(minTimestamp) {
+			minTimestamp = acc.accumulatorData[i][acc.firstMapKeysOfAccumulatorData[i]][0].GetEventTimestamp()
+			index = i
+		}
+	}
+	acc.createNewMap(event, index)
+	return index
+}
+
+func (acc *Accumulator) findIndexByTimestamp(event *evData.EventData) int {
+	for i := range acc.accumulatorData {
+		if len(acc.accumulatorData[i]) == 0 {
+			acc.createNewMap(event, i)
+			return i
+		}
+		firstKey := acc.firstMapKeysOfAccumulatorData[i]
```

Comment:
[Required]: **DO NOT ACCESS A SLICE BEFORE CHECKING IT'S LENGTH!!!!!!!!!!!!!!!!!!**
I do not want to see such code again!